### PR TITLE
QLoRA with Llama 3.1 405B

### PR DESCRIPTION
**Pull request by pbontrager**
**Edit**: this PR has been resuscitated. The main change since the previous version is adding a new way to pass checkpoint files. Now we can just do 

```
  checkpoint_files:
    formatted_string: model-{}-of-{}.safetensors
    max_filename: 00191
```

without having to copy-paste 191 rows of config. We should integrate this into all our 70B configs too when we get a chance.

## Test plan

Added some new tests for the `FormattedCheckpointFiles` class.

```
pytest tests/torchtune/training/checkpointing/test_checkpointer_utils.py
...
====== 11 passed, 1 warning in 0.14s ===================
```

E2E test:

```
tune run --nproc_per_node 8 lora_finetune_distributed --config llama3_1/405B_qlora \
max_steps_per_epoch=2 gradient_accumulation_steps=2 epochs=2
...
1|2|Loss: 1.7182660102844238: 100%|████████████████████████████| 2/2 [02:39<00:00, 79.59s/it]saving checkpoint
INFO:torchtune.utils._logging:Adapter checkpoint of size 1.25 GB saved to /tmp/Meta-Llama-3.1-405B-Instruct/adapter_0.pt
INFO:torchtune.utils._logging:Adapter checkpoint of size 1.25 GB saved to /tmp/Meta-Llama-3.1-405B-Instruct/adapter_model.bin
INFO:torchtune.utils._logging:Adapter checkpoint of size 0.00 GB saved to /tmp/Meta-Llama-3.1-405B-Instruct/adapter_config.json
INFO:torchtune.utils._logging:Recipe checkpoint of size 2.50 GB saved to /tmp/Meta-Llama-3.1-405B-Instruct/recipe_state.pt
1|2|Loss: 1.7182660102844238: 100%|██████████████████████████████| 2/2 [02:53<00:00, 86.50s/it]
  0%|                                                                                                                                                                                                                                                                     | 0/2 [00:00<?, ?it/ssaving checkpoint971073150635: 100%|████████████████████████| 2/2 [02:38<00:00, 79.58s/it]
INFO:torchtune.utils._logging:Adapter checkpoint of size 1.25 GB saved to /tmp/Meta-Llama-3.1-405B-Instruct/adapter_1.pt
INFO:torchtune.utils._logging:Adapter checkpoint of size 1.25 GB saved to /tmp/Meta-Llama-3.1-405B-Instruct/adapter_model.bin
INFO:torchtune.utils._logging:Adapter checkpoint of size 0.00 GB saved to /tmp/Meta-Llama-3.1-405B-Instruct/adapter_config.json
INFO:torchtune.utils._logging:Saving final epoch checkpoint.
INFO:torchtune.utils._logging:Please note that you have set adapter_only=True, so only adapter weights will be saved.You need to merge the adapter weights into your base model for further use. See FullModelHFCheckpointer.save_checkpoint for more details.
2|4|Loss: 2.5931971073150635: 100%|███████████████████████████████████| 2/2 [02:45<00:00, 82.55s/it]
```

Setting aside the fact that our logs are annoying here, this works.

#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

This PR adds a Lllama 405B QLoRA config for the lora_finetune_fsdp2 recipe. This config requires 8 x A100s to run. The config successfully finetunes Llama 405B on the alpaca dataset as an example. There are several caveats for getting this large model to fit on a single node:
- Model checkpointing causes NCCL timeout errors, so we only save the adapter weights that can be merged afterward
- Training is slow, at around 10 minutes for 16 gradient steps (this should improve with compile support)
- Can only fit < 4k context length

This PR was jointly made with me and @joecummings 

#### Changelog
- New config
- New model builders
- Updated Docs for new model builders
- Updated fsdp2 recipe save checkpoint to support only saving adapter weights

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
	- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

Test Memory + Checkpointing
`tune run --nproc_per_node 8 lora_finetune_fsdp2 --config llama3_1/405B_qlora log_peak_memory_stats=True dataset.packed=True dataset.max_seq_len=2048 max_steps_per_epoch=2 epochs=2`

Test Model Loss
`tune run --nproc_per_node 8 lora_finetune_fsdp2 --config llama3_1/405B_qlora`

### CODE REVIEW

File: torchtune/models/llama3_1/_model_builders.py [hunk 1]
```diff
@@ -63,6 +63,27 @@ def llama3_1_70b() -> TransformerDecoder:
     )
 
 
+def llama3_1_405b() -> TransformerDecoder:
+    """
+    Builder for creating a Llama3 model initialized w/ the default 405B parameter values.
```
Comment (SalmanMohammadi) [inline torchtune/models/llama3_1/_model_builders.py:68]: nit: Llama3.1
here and in the return

File: torchtune/utils/_distributed.py [hunk 2]
```diff
@@ -361,12 +361,17 @@ def load_from_full_model_state_dict(
...
+    trainable_only: bool = False,
 ) -> Dict[str, Any]:
     """
     Converting sharded state dict into a full state dict on cpu
     Returning non-empty result on rank0 to avoid peaking cpu memory
+
+    Args:
+        model (FSDPModule): wrapped module
+        is_rank_zero (bool): flag to check if the process is on rank 0
+        trainable_only (bool): flag to only return state dict of trainable parameters
```
[the remark this answers is not in the document]
Comment (pbontrager) [reply torchtune/utils/_distributed.py:373]: Initially I thought just saving the adapter would be enough, but for the 405B model, consolidating the state_dict is also prohibitively expensive. This update can be added to the other recipes after this lands.

**Comment by SalmanMohammadi** [comment]
Couple of small comments (looks like CI is failing too?). This is huge. 

Two points as an an aside:

- I see this is a dev recipe right? What would bring it into the stable recipes?

- Following from above, would something like this would warrant a writeup to help get the good word out about the amazing work you guys have been doing here, and how torchtune enables feats like this?

**Comment by joecummings** [comment]
> * I see this is a dev recipe right? What would bring it into the stable recipes?

The only reason this is dev is that FSDP2 support just dropped in PyTorch stable a few days ago, so we want to play around with it more before moving to stable. We need FSDP2 support b/c it plays nicely with QLoRA. 

> * Following from above, would something like this would warrant a writeup to help get the good word out about the amazing work you guys have been doing here, and how torchtune enables feats like this?

We'd love this - I can definitely promote on Discord and Twitter and _start_ on a formal blog writeup, but would appreciate any help here :)

File: torchtune/utils/_distributed.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joecummings) [reply torchtune/utils/_distributed.py:373]: No NCCL timeouts?

**Comment by SalmanMohammadi** [comment]
> We'd love this - I can definitely promote on Discord and Twitter and _start_ on a formal blog writeup, but would appreciate any help here :)

I'd love to! Happy to help out here in any capacity - reviewing to help fleshing out a rough sketch.

File: torchtune/models/llama3_1/__init__.py [hunk 3]
```diff
@@ -7,10 +7,13 @@
 from ._component_builders import llama3_1, lora_llama3_1
```
Comment (SalmanMohammadi) [inline torchtune/models/llama3_1/__init__.py:7]: linter unhappy : ( need to add to `__all__`?

Comment (pbontrager) [reply torchtune/models/llama3_1/__init__.py:7]: It's in __all__, no idea what's triggering this. The pre-commit flake8 passes too

File: torchtune/models/llama3_1/__init__.py [hunk 4]
```diff
@@ -19,9 +22,12 @@
     "llama3_1",
     "llama3_1_8b",
     "llama3_1_70b",
+    "llama3_1_405b"
     "lora_llama3_1",
     "lora_llama3_1_8b",
     "lora_llama3_1_70b",
+    "lora_llama3_1_405b"
```
Comment (SalmanMohammadi) [inline torchtune/models/llama3_1/__init__.py:29]: also here

File: torchtune/models/llama3_1/__init__.py [hunk 5]
```diff
@@ -19,9 +22,12 @@
     "llama3_1",
     "llama3_1_8b",
     "llama3_1_70b",
+    "llama3_1_405b"
```
[the remark this answers is not in the document]
Comment (pbontrager) [reply torchtune/models/llama3_1/__init__.py:25]: :face_palm:

File: recipes/configs/dev/llama3_1/405B_qlora_fsdp2.yaml [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+  lora_alpha: 32
+
+tokenizer:
+  _component_: torchtune.models.llama3.llama3_tokenizer
+  path: /tmp/Meta-Llama-3.1-405B-Instruct/original/tokenizer.model
+
+checkpointer:
+  _component_: torchtune.utils.FullModelHFCheckpointer
+  checkpoint_dir: /tmp/Meta-Llama-3.1-405B-Instruct/
+  checkpoint_files: [
```
Comment (ebsmothers) [inline recipes/configs/dev/llama3_1/405B_qlora_fsdp2.yaml:32]: Guys I think the time has come.. we need to refactor `checkpoint_files` field to optionally support the combination of (a) number of files and (b) strf filename format (or something like that). This is like 5 minutes of work and will save us from abominations like this

**Comment by SalmanMohammadi** [comment]
```bash
pydoclint................................................................Failed
- hook id: pydoclint
- exit code: 1

Skipping files that match this pattern: tests/torchtune/models/(\w+)/scripts/
recipes/dev/lora_finetune_fsdp2.py
torchtune/_recipe_registry.py
torchtune/models/llama3_1/__init__.py
torchtune/utils/_checkpointing/_checkpointer.py
torchtune/utils/_distributed.py
torchtune/utils/_distributed.py
    361: DOC201: Function `get_full_model_state_dict` does not have a return section in docstring
    361: DOC501: Function `get_full_model_state_dict` has "raise" statements, but the docstring does not have a "Raises" section
Loading config from user-specified .toml file: pyproject.toml
Found options defined in pyproject.toml:
{'style': 'google', 'check_return_types': 'False', 'exclude': 'tests/torchtune/models/(\\w+)/scripts/'}
```

welcome back to hell, brother

**Comment by joecummings** [comment]
Currently hitting this error with compile, no speedups:

```
[rank1]:W0730 05:54:09.440000 86176 torch/_dynamo/convert_frame.py:795] [8/8] torch._dynamo hit config.cache_size_limit (8)
[rank1]:W0730 05:54:09.440000 86176 torch/_dynamo/convert_frame.py:795] [8/8]    function: 'forward' (/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:146)
[rank1]:W0730 05:54:09.440000 86176 torch/_dynamo/convert_frame.py:795] [8/8]    last reason: 8/0: ___check_type_id(L['self'], 206876688)
[rank1]:W0730 05:54:09.440000 86176 torch/_dynamo/convert_frame.py:795] [8/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
[rank1]:W0730 05:54:09.440000 86176 torch/_dynamo/convert_frame.py:795] [8/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
[rank4]:W0730 05:54:09.441000 86179 torch/_dynamo/convert_frame.py:795] [8/8] torch._dynamo hit config.cache_size_limit (8)
[rank4]:W0730 05:54:09.441000 86179 torch/_dynamo/convert_frame.py:795] [8/8]    function: 'forward' (/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:146)
[rank4]:W0730 05:54:09.441000 86179 torch/_dynamo/convert_frame.py:795] [8/8]    last reason: 8/0: ___check_type_id(L['self'], 206881776)
[rank4]:W0730 05:54:09.441000 86179 torch/_dynamo/convert_frame.py:795] [8/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
[rank4]:W0730 05:54:09.441000 86179 torch/_dynamo/convert_frame.py:795] [8/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
[rank7]:W0730 05:54:09.441000 86182 torch/_dynamo/convert_frame.py:795] [8/8] torch._dynamo hit config.cache_size_limit (8)
[rank7]:W0730 05:54:09.441000 86182 torch/_dynamo/convert_frame.py:795] [8/8]    function: 'forward' (/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:146)
[rank7]:W0730 05:54:09.441000 86182 torch/_dynamo/convert_frame.py:795] [8/8]    last reason: 8/0: ___check_type_id(L['self'], 206893040)
[rank7]:W0730 05:54:09.441000 86182 torch/_dynamo/convert_frame.py:795] [8/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
[rank7]:W0730 05:54:09.441000 86182 torch/_dynamo/convert_frame.py:795] [8/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
[rank0]:W0730 05:54:09.441000 86175 torch/_dynamo/convert_frame.py:795] [8/8] torch._dynamo hit config.cache_size_limit (8)
[rank0]:W0730 05:54:09.441000 86175 torch/_dynamo/convert_frame.py:795] [8/8]    function: 'forward' (/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:146)
[rank0]:W0730 05:54:09.441000 86175 torch/_dynamo/convert_frame.py:795] [8/8]    last reason: 8/0: ___check_type_id(L['self'], 212654320)
[rank0]:W0730 05:54:09.441000 86175 torch/_dynamo/convert_frame.py:795] [8/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
[rank0]:W0730 05:54:09.441000 86175 torch/_dynamo/convert_frame.py:795] [8/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
[rank6]:W0730 05:54:09.441000 86181 torch/_dynamo/convert_frame.py:795] [8/8] torch._dynamo hit config.cache_size_limit (8)
[rank6]:W0730 05:54:09.441000 86181 torch/_dynamo/convert_frame.py:795] [8/8]    function: 'forward' (/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:146)
[rank6]:W0730 05:54:09.441000 86181 torch/_dynamo/convert_frame.py:795] [8/8]    last reason: 8/0: ___check_type_id(L['self'], 206868944)
[rank6]:W0730 05:54:09.441000 86181 torch/_dynamo/convert_frame.py:795] [8/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
[rank6]:W0730 05:54:09.441000 86181 torch/_dynamo/convert_frame.py:795] [8/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
[rank3]:W0730 05:54:09.442000 86178 torch/_dynamo/convert_frame.py:795] [8/8] torch._dynamo hit config.cache_size_limit (8)
[rank3]:W0730 05:54:09.442000 86178 torch/_dynamo/convert_frame.py:795] [8/8]    function: 'forward' (/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:146)
[rank3]:W0730 05:54:09.442000 86178 torch/_dynamo/convert_frame.py:795] [8/8]    last reason: 8/0: ___check_type_id(L['self'], 206881856)
[rank3]:W0730 05:54:09.442000 86178 torch/_dynamo/convert_frame.py:795] [8/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
[rank3]:W0730 05:54:09.442000 86178 torch/_dynamo/convert_frame.py:795] [8/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
[rank2]:W0730 05:54:09.442000 86177 torch/_dynamo/convert_frame.py:795] [8/8] torch._dynamo hit config.cache_size_limit (8)
[rank2]:W0730 05:54:09.442000 86177 torch/_dynamo/convert_frame.py:795] [8/8]    function: 'forward' (/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:146)
[rank2]:W0730 05:54:09.442000 86177 torch/_dynamo/convert_frame.py:795] [8/8]    last reason: 8/0: ___check_type_id(L['self'], 206884400)
[rank2]:W0730 05:54:09.442000 86177 torch/_dynamo/convert_frame.py:795] [8/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
[rank2]:W0730 05:54:09.442000 86177 torch/_dynamo/convert_frame.py:795] [8/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
[rank5]:W0730 05:54:09.443000 86180 torch/_dynamo/convert_frame.py:795] [8/8] torch._dynamo hit config.cache_size_limit (8)
[rank5]:W0730 05:54:09.443000 86180 torch/_dynamo/convert_frame.py:795] [8/8]    function: 'forward' (/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:146)
[rank5]:W0730 05:54:09.443000 86180 torch/_dynamo/convert_frame.py:795] [8/8]    last reason: 8/0: ___check_type_id(L['self'], 206877312)
[rank5]:W0730 05:54:09.443000 86180 torch/_dynamo/convert_frame.py:795] [8/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
[rank5]:W0730 05:54:09.443000 86180 torch/_dynamo/convert_frame.py:795] [8/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
```

**Comment by bdhirsh** [comment]
@joecummings it's a bit hard to tell exactly what we're recompiling on from the logs (since it just includes a type id). But one known cause of recompilations using DTensor today is that we overspecialize on all of the DTensor metadata (`Placement`/`DeviceMesh`).

This PR fixes it, so once it lands can you try again? (and we can stare at the recompiles some more if that doesn't fix it) https://github.com/pytorch/pytorch/pull/124401

File: torchtune/training/checkpointing/_utils.py
Comment (joecummings) [inline torchtune/training/checkpointing/_utils.py:1]: test???????!?!??!?!?!?!

Comment (joecummings) [reply torchtune/training/checkpointing/_utils.py:1]: Also what the hell does `strf_name` stand for?

**Review by joecummings — APPROVED** [review]
Why not allow a regex pattern for files

File: torchtune/training/checkpointing/_utils.py
Comment (ebsmothers) [reply torchtune/training/checkpointing/_utils.py:1]: string formatted name. Open to better names though

Comment (ebsmothers) [reply torchtune/training/checkpointing/_utils.py:1]: > test???????!?!??!?!?!?!

ok ok ok. The `FormattedCheckpointFiles` class was hacked together pretty quickly, first just wanted to make sure it made sense to people

**Pull request merged**